### ISSUE TITLE
Elixir bar on every game mode

### ISSUE THREAD

**Issue by warklp** [AmarSaini/Clash-Royale-AI-Card-Tracker#2]
All game mode now show the elixir bar, that means, isnt working anymore, telling wrong cards and not working couting elixir, you know how to fix it?

**Comment by AmarSaini** [comment]
It’s because Supercell updated the layout of the spectator mode. The AI isn’t looking at the right spots for the card locations. It needs to be updated to look at the right areas.

**Comment by warklp** [comment]
I fixed the layout, but i am having problem with the last card of the deck, maybe you can help me
![image](https://user-images.githubusercontent.com/49003984/55127699-d310a500-50ef-11e9-9fd3-53376b2e2ed4.png)

**Comment by AmarSaini** [comment]
Check where the files are being cropped and saved at. Stop the program while it’s running to see how the 8th card is getting cropped.

Amar Saini

M.S. - Electrical Engineering & Computer Science
School of Engineering, University of California, Merced
B.S. - Computer Science & Engineering
School of Engineering, University of California, Merced

________________________________

**Comment by AmarSaini** [comment]
It looks like you have the testCNN.png open, close the image and try again. (Permission denied to write since the file is open).

**Comment by warklp** [comment]
i'll see that, and another thing, sometimes i get this error and the script closes 
![image](https://user-images.githubusercontent.com/49003984/55128382-1f5ce480-50f2-11e9-9281-fd3b6c23620f.png)

**Comment by AmarSaini** [comment]
Interesting, do you have any other cases of this happening?

**Comment by warklp** [comment]
When i start and my opponent already played 1 card and then he play that card again this always happens

**Comment by AmarSaini** [comment]
I want to say it has to do with the formatting still. Can you run the program and then close out of it during the game. Afterwards, go to your "testData" folder. This folder is where temporary crops of each of the 8 cards are saved. Maybe they aren't getting cropped right, check to see if the cropped images are indeed the cards.

**Comment by warklp** [comment]
Another error just happend, everything was fine, game was already on overtime..
![image](https://user-images.githubusercontent.com/49003984/55129299-54b70180-50f5-11e9-8051-bf23ac3597eb.png)

**Comment by AmarSaini** [comment]
It looks like it's an error in the save function for the PIL library, give me a moment

**Comment by warklp** [comment]
Seems like its cropped just fine, but stop working for no reason 
![image](https://user-images.githubusercontent.com/49003984/55129549-487f7400-50f6-11e9-9f1a-97db6bfda6dd.png)

**Comment by AmarSaini** [comment]
I see the problem here, try placing a r in front of the file path for any open/save function calls, and maybe use the absolute address.
Check this out: https://stackoverflow.com/questions/39737448/random-errors-when-using-pil-on-windows
"[Errno 22] invalid argument" usually relates to issues regarding file paths

**Comment by warklp** [comment]
all the cards cropped just fine and i get that error, saying that a card isnt in the list
![image](https://user-images.githubusercontent.com/49003984/55130116-677f0580-50f8-11e9-80d4-0126c288b273.png)

**Comment by AmarSaini** [comment]
As for this error, the opponent's hand isn't getting populated properly. Mind checking how the testData2 folder looks? It should be the areas slightly above the 8 cards.

**Comment by warklp** [comment]
it is above the 8 cards, but the layout now its different, maybe a little bit confuse for the script 
![image](https://user-images.githubusercontent.com/49003984/55130331-3d7a1300-50f9-11e9-8962-a268fc54ce18.png)

**Comment by AmarSaini** [comment]
I see the problem now. The old spectator background/layout was just brown, and it didn't show the elixir count. But now the layout has changed, it shows the current elixir in the brown background area. This makes the neural network think that cards are being placed when they really aren't, because it sees pink which was the old way of determining when a card was played. I would need to retrain this neutral network for it to have correct predictions. This would require a new set of training data, and it may require much more data to train, or perhaps a stronger training method (since the predictions are now harder: pink background with pink pop-ups).

**Comment by warklp** [comment]
True, that's probably why, thank you, i'll try to do something about it.

**Comment by AmarSaini** [comment]
Let me know if you think of anything! I see 2 options as of now, 1) retrain the 2nd neural network with new training data, or 2) find a new way to determine played cards.

**Comment by warklp** [comment]
Well, ive noticed that when the opponent play a card the card go 1 or 2 cm up for a second or less, maybe thats a opening for another way to determine cards played.

**Comment by AmarSaini** [comment]
That’s true, but that means your computer will have to be able to perform predictions a second each. I believe I choose the pink icon because it lasted longer. It’s worth a try though!

Amar Saini

M.S. - Electrical Engineering & Computer Science
School of Engineering, University of California, Merced
B.S. - Computer Science & Engineering
School of Engineering, University of California, Merced

________________________________

**Comment by warklp** [comment]
Another thing ive noticed, most of the times que program dont recognize the "golden cards" just like the last image i shared, do you have any idea of how to make the program recognize the 2 types of cards? golden and not-golden?

**Comment by AmarSaini** [comment]
Network would also need to be retained for that. Or maybe crop inside the card image (get rid of the gold border) that may work

Amar Saini

M.S. - Electrical Engineering & Computer Science
School of Engineering, University of California, Merced

B.S. - Computer Science & Engineering
School of Engineering, University of California, Merced

________________________________

**Comment by warklp** [comment]
Hello again, I'am still trying to fix the invalid argument issue, i don't know where to put the r, can you tell me in which lines?

**Comment by AmarSaini** [comment]
Hey,

Line 661 should be one. Anywhere you are accessing images for reading/writing. CTRL-F "imsave" for a few.

Amar Saini

M.S. - Electrical Engineering & Computer Science
School of Engineering, University of California, Merced

B.S. - Computer Science & Engineering
School of Engineering, University of California, Merced

________________________________

**Comment by ComputerApp1234** [comment]
Any chance for a fix by you, because of the most recent updates or was it just a personal project related to neural networks?

**Comment by ComputerApp1234** [comment]
Any chance for a fix by you, because of the most recent updates or was it just a personal project to deal with neural networks?

**Comment by AmarSaini** [comment]
Hey, @ComputerApp1234. Yea I noticed the game has updated, such as new spectator layout (different location of cards), elixir bar now shows while spectating if I remember correctly, new cards/some cards have been redesigned. I haven't been working on it too much, but I may make a revamped version that uses a different framework, such as PyTorch or even fast.ai (definitely go check them out!). Also, I may make it so you don't need to be using 1920x1080 specifically (make the card locations depending on max width/height, rather than hard-coded values of a 1920x1080 screen). I'll make an announcement/update to read me soon!

**Comment by ComputerApp1234** [comment]
Thank you very much for the fast reply. Like this it would be possible to use other emulators instead of BlueStacks?
And an update by you would be great!

**Comment by AmarSaini** [comment]
I'm not too sure about any other emulators, are you familiar with any?

**Comment by ComputerApp1234** [comment]
I usually use MEmu to play Clash Royale.

**Comment by AmarSaini** [comment]
Ah interesting! Have you tried bluestacks as well? Which do you prefer/is more popular?

**Comment by ComputerApp1234** [comment]
I've tried both, but i prefer MEmu, because it feels more stable in my opinion. But this could also be because of my settings. The main reason is i'm more used to MEmu.

**Comment by AmarSaini** [comment]
Interesting I'll check it out
